platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java

```diff
@@ -1305,23 +1394,58 @@ private static <K, V> void purge(final ConcurrentArray<Mutation<K, V>> array, fi
      * obsolete, if there is a newer mutation for the same key.
      *
-     * @param array
+     * @param array the list of mutations to process
+     * @param index the corresponding index, it's used to look up the newest mutations
+     *              for a key
+     * @param newestVersion the newest version of all mutations in the array
      * @param <K>
      * 		The key type used in the index
      * @param <V>
      * 		The value type referenced by the mutation list
      */
-    private static <K, V> void filterMutations(final ConcurrentArray<Mutation<K, V>> array) {
+    private static <K, V> void filterMutations(
+            final ConcurrentArray<Mutation<K, V>> array,
+            final Map<K, Mutation<K, V>> index,
+            final long newestVersion,
+            final AtomicLong filteredCounter) {
+        filteredCounter.set(0);
         final Consumer<Mutation<K, V>> action = mutation -> {
             // local variable is required because mutation.next can be changed by another thread to null
             // see https://github.com/hashgraph/hedera-services/issues/7046 for the context
             Mutation<K, V> nextMutation = mutation.next;
             if (nextMutation != null) {
+                assert !nextMutation.isFiltered();
                 nextMutation.setFiltered();
+                filteredCounter.incrementAndGet();
+                if (nextMutation.isNew()) {
+                    // nextMutation is to put a new element into a virtual map. The element doesn't
+                    // exist in the data source. If this mutation is filtered, there must be a newer
+                    // mutation for the same key. If that newer mutation has the "deleted" flag, the
+                    // element should never be flushed to disk
+                    final Mutation<K, V> latestMutation = index.get(nextMutation.key);
+                    assert latestMutation != null;
+                    final Mutation<K, V> latestMutationUpToVersion = lookup(latestMutation, newestVersion);
+                    assert latestMutationUpToVersion != null;
+                    if (latestMutationUpToVersion.isDeleted()) {
+                        assert !latestMutationUpToVersion.isFiltered();
+                        // If the latest mutation up to newestVersion is "deleted", and there are no
+                        // newer mutations, the whole entry for the key can be removed from the index.
+                        // It's safe to do so here, as there are no references to copies older than
+                        // newestVersion and there are no mutations in versions newer than newestVersion
+                        if (latestMutation == latestMutationUpToVersion) {
+                            latestMutationUpToVersion.setFiltered();
+                            filteredCounter.incrementAndGet();
+                            index.remove(latestMutation.key);
```

Comment:
Minor q: if there're mutations for the same key in versions greater than the `newestVersion`, would removing the key from the index affect those later versions in any way?

Reply:
All these mutations are effectively just dropped.

Here is an example. Assume for a certain path, the entry in the concurrent map (`index` param for this method) contains a linked list of mutations like this:

```
path: (version3, deleted, null value) -> (version2, value2) -> (version1, new, value1) -> null
```

value1 was inserted in version1, then updated in version2, then removed in version 3. The concurrent array will contain these three mutations, not necessarily next to each other. When mutation for version2 is processed, the next mutation in the list (for version1) is marked as filtered. Since this filtered mutation is "new", the first mutation in the whole linked list is checked, for version3. It's marked as "deleted", and therefore mutation3 is filtered, and the map entry for this path is removed. Some time later (or earlier), mutation3 is processed, it will filter out mutation2, this is fine.

Reply:
Sounds good.